### ISSUE TITLE
DataLogWriterNotPrimaryException thrown indefinitely

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#4654]
**Problem description**
We have observe this issue 2 times already, in different experiments (induced instance failures, benchmark workload). The problem is as follows:
1) A Segment Store owns several containers:
``` 
2020-03-29 19:23:04,068 27314400 [core-27] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Container Changes: Desired = [20, 21, 22, 23, 12, 13, 14, 15], Current = [20, 21, 22, 23, 12, 13, 14, 15], PendingTasks = [], ToStart = [], ToStop = [].
```
2) At some point, it looks like there is a network glitch, given that few Segment Stores have problem to publish metrics to Influx during a short period of time:
```
2020-03-29 19:26:41,453 27531785 [influx-metrics-publisher] ERROR i.m.influx.InfluxMeterRegistry - failed to send metrics to influx
...
2020-03-29 19:33:30,469 27940801 [influx-metrics-publisher] ERROR i.m.influx.InfluxMeterRegistry - failed to send metrics to influx
```
After that metrics seem to be reported correctly, which discards that the problem is just about the InfluxDB instance.

3) During this period of time, we have to note that the system is under high load, and the Segment Store is throttling on most of its containers as Bookkeeper seems to be saturated:
```
2020-03-29 19:27:25,148 27575480 [core-16] WARN  i.p.s.server.logs.Throttler - Throttler[23]: Processing delay = 25000ms (Max=true, Reason=DurableDataLog).
2020-03-29 19:27:25,150 27575482 [core-24] WARN  i.p.s.server.logs.Throttler - Throttler[23]: Processing delay = 25000ms (Max=true, Reason=DurableDataLog).
2020-03-29 19:27:25,150 27575482 [core-10] WARN  i.p.s.server.logs.Throttler - Throttler[20]: Processing delay = 18094ms (Max=false, Reason=DurableDataLog).
2020-03-29 19:27:25,151 27575483 [core-20] WARN  i.p.s.server.logs.Throttler - Throttler[23]: Processing delay = 25000ms (Max=true, Reason=DurableDataLog).
```

4) In this situation, we observe the `DataLogWriterNotPrimaryException` in several containers:
```
2020-03-29 19:30:05,136 27735468 [core-10] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[20]: Iteration[845188].Error.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/pravega/segmentstore/containers/0/2/20').
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.persistMetadata(BookKeeperLog.java:857)
```
```
2020-03-29 19:30:05,136 27735468 [core-29] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[13]: Iteration[774304].Error.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/pravega/segmentstore/containers/3/1/13').
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.persistMetadata(BookKeeperLog.java:857)
```
```
2020-03-29 19:30:05,136 27735468 [core-18] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[22]: Iteration[870488].Error.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/pravega/segmentstore/containers/2/2/22').
	at io.pravega.segmentstore.storage.impl.bookkeeper.BookKeeperLog.persistMetadata(BookKeeperLog.java:857)
```

5) These containers are left throwing `DataLogWriterNotPrimaryException` indefinitely.

**Problem location**
Possibly at `BookKeeperLog.java`

**Suggestions for an improvement**
Fix the issue to make the `StorageWriter` to recover eventually.

**Comment by RaulGracia** [comment]
Regarding point 2), I observed that this problem happens in 2 different Segment Stores, and in both cases it is preceded by a disconnection from Zookeeper:
```
(Segment Store 1)

2020-03-29 19:29:58,091 27728423 [main-SendThread(example-client:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6666ms for sessionid 0x20003ca2f8dd6c2
2020-03-29 19:29:58,091 27728423 [main-SendThread(example-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6666ms for sessionid 0x20003ca2f8dd6c2, closing socket connection and attempting reconnect
...
2020-03-29 19:30:05,136 27735468 [core-10] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[20]: Iteration[845188].Error.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/pravega/segmentstore/containers/0/2/20').
```

```
(Segment Store 2)

2020-03-29 19:29:59,708 27658476 [main-SendThread(example-client:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6667ms for sessionid 0x20003ca2f8dd6d7
2020-03-29 19:29:59,709 27658477 [main-SendThread(example-client:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6667ms for sessionid 0x20003ca2f8dd6d7, closing socket connection and attempting reconnect
...
2020-03-29 19:30:04,856 27663624 [core-13] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[7]: Iteration[864164].Error.
io.pravega.segmentstore.storage.DataLogWriterNotPrimaryException: Unable to acquire exclusive write lock for log (path = 'pravega/pravega/segmentstore/containers/7/0/7'
```

**Comment by RaulGracia** [comment]
Closed via #4666.
